### DISCUSSION TITLE
Unlink materials of instanced asset?

### DISCUSSION THREAD

**Discussion by jimbz** [google/filament#3992] [Q&A] [answered]
Hi,

Thank you for the good work in making Filament powerful and lightweight.

I have an asset with many instances (`AssetLoader.createInstancedAsset`), and I realized that all the assets share the same material instance. Is is possible to "unlink" those instances so that their parameters can be updated independently?

I tried something like this:
```kotlin
                    val material = renderableManager.getMaterialInstanceAt(
                        renderableInstance,
                        primitiveIndex
                    ).material

                    val materialInstance = material.createInstance()

                    renderableManager.setMaterialInstanceAt(
                        renderableInstance,
                        primitiveIndex,
                        materialInstance
                    )
```
which builds and runs but does not seem to preserve the original parameters (the materials become invisible).

It is possible I did something wrong but I would like confirmation that my mental model of the materials and material instances is correct in the context of an instanced assets.

Thank you!

**Comment by romainguy** [comment, marked as answer]
What you did is correct, unfortunately an instance only gets its default parameters from the material, and it seems like we only set parameters on instances in `gltfio`.

@prideout Do you think we could set default parameters on materials for `gltfio`? Not clear if it makes sense or not but it sound like it could be useful for cases like this.

**Reply by jimbz** [reply]
Thanks for the quick reply! Glad that I wasn't too far off.

I am open to any hacks to try and enable this use case, including fiddling in Filament code.  If there was a `getParameter` I could try and manually copy the parameters to the new instances, but that doesn't seem to be exposed either...

**Reply by romainguy** [reply]
There's a PR pending to add `getParameter()` methods but it's not great as you need to know the list of parameters and their types (we can expose this as well if we don't already). Setting defaults seems much easier for your use case though :)

**Reply by prideout** [reply]
My understanding is that Jim needs to set up unique parameter values to each asset instance, so I don't understand how setting default parameters on the base material would help.

I think the code in the original post is correct, so it would be good to figure out why the mesh is disappearing.  One theory is that this is due to asynchronous loading functionality, which inspects material instances to figure out when to un-hide each mesh.

@jimbz  are you using `loadResources` or `asyncBeginLoad`?

**Reply by jimbz** [reply]
@prideout I need to change only a single parameter, while preserving the other ones. In this case it's actually a new opacity parameter that I added locally inside gltfio but I could see use cases with changing existing parameters like the color for example.

I am using `asyncBeginLoad` but my models are not texture-heavy so I think I could easily switch to `loadResources`.

**Reply by prideout** [reply]
As an experiment you can try switching to `loadResources` to see if that fixes your disappearing mesh issue.  The big comment block in `AssetLoader.java` shows how to do this.  (note the use of `addEntities` rather than `popRenderables`)

**Reply by jimbz** [reply]
No luck... So are you saying that the default parameters are indeed set on the `gltfio` materials so creating and using a new instance like I did, without changing any parameter should in theory look the same?

**Reply by prideout** [reply]
Well you'd still need to set up every parameter.  I'm starting to see what Romain was saying now, especially since it sounds like your your mesh is disappearing due to opacity.

I'm just not sure that setting up default parameters are what you want, since gltfio can use the same base material to create many different material instances, each of which is customized by various glTF nodes.  Seems like you really need the `getParameter` stuff that Romain was alluding to, or a new `clone()` method proffered by MaterialInstance.

**Reply by jimbz** [reply]
Gotcha, the materials are reused between assets which is why the default parameters do not make sense indeed. I will look into cloning instances using the getParameter PR.

Thank you for taking the time to walk me through the situation!
